### PR TITLE
Feature: basic auth for API v3

### PR DESCRIPTION
**Pull request by machisuji**
WP [#19294](https://community.openproject.org/work_packages/19294)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://machisuji-feature-basic-auth.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://machisuji-feature-basic-auth.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: lib/api/root.rb [hunk 1]
```diff
@@ -59,19 +63,32 @@ def call(object, _env)
...
+        warden.authenticate! scope: API_V3
+
+        User.current = warden.user scope: API_V3
+
+        if Setting.login_required? && not_logged_in?
           raise API::Errors::Unauthenticated
         end
       end
 
+      def not_logged_in?
```
Comment (NobodysNightmare) [inline lib/api/root.rb]: Can we please avoid using negated bool-methods? (when there is no existing positive method)

Soon someone will write `!not_logged_in?`, because the `not_logged_in?` method was already available...

File: lib/api/root.rb [hunk 2]
```diff
@@ -59,19 +63,32 @@ def call(object, _env)
...
+        # An admin SystemUser is anonymous but still a valid user to be logged in.
```
Comment (NobodysNightmare) [inline lib/api/root.rb:89]: :+1: (for clarifying comment)

File: lib/api/root.rb [hunk 3]
```diff
@@ -59,19 +63,32 @@ def call(object, _env)
...
+
+      def warden
+        env['warden']
       end
 
       def authenticate
-        if Setting.login_required? && (current_user.nil? || current_user.anonymous?)
+        warden.authenticate! scope: API_V3
+
+        User.current = warden.user scope: API_V3
```
Comment (NobodysNightmare) [inline lib/api/root.rb:81]: That means you moved the side-effect "sets `User.current`" from `current_user` to `authenticate`, right?

This sounds reasonable, but you should update this: https://github.com/machisuji/openproject/blob/846e331c784fcc2a9ae1ad42cae3adf65a273454/lib/api/root.rb#L165-L173

**Comment by machisuji** [comment]
@NobodysNightmare Made logged_in? logic positive and removed the obsolete comment/code in before.

File: config/initializers/warden.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+require 'open_project/authentication'
+
+# Strategies provided by OpenProject:
+require 'open_project/authentication/strategies/warden/basic_auth_failure'
+require 'open_project/authentication/strategies/warden/global_basic_auth'
+require 'open_project/authentication/strategies/warden/user_basic_auth'
+require 'open_project/authentication/strategies/warden/session'
+
+strategies = [
```
Comment (NobodysNightmare) [inline config/initializers/warden.rb]: WHY U NOT USE A DICTIONARY?

File: lib/open_project/authentication.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  module Authentication
+    class << self
+      ##
+      # Updates the used warden strategies for a given scope. The strategies will be tried
+      # in the order they are set here. Plugins can call this to add or remove strategies.
+      # For available scopes please refer to `OpenProject::Authentication::Scope`.
+      #
+      # @param [Symbol] scope The scope for which to update the used warden strategies.
+      # @param [Boolean] store Indicates whether the user should be stored in the session
+      #                        for this scope.
```
Comment (NobodysNightmare) [inline lib/open_project/authentication.rb]: What happens if it is not provided?

Without looking deeper down the rabbit hole I can't tell wheter `nil` means true, false or whatever it might mean...

Comment (machisuji) [inline lib/open_project/authentication.rb]: The `store_defaults` value for that scope will remain unchanged which means what ever default value is [set by the Manager](https://github.com/machisuji/openproject/blob/feature/basic_auth/lib/open_project/authentication/manager.rb#L28) (`false`) will be used.

Comment (machisuji) [inline lib/open_project/authentication.rb]: The point is that as a plugin developer I may just want to touch the strategies (add one, remove one, reorder) while not caring about the persistence of the logged in user. Do you want me to mention that in the comment?

File: lib/open_project/authentication.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        current_strategies = Array(Manager.scope_strategies[scope])
+
+        Manager.store_defaults[scope] = store unless store.nil?
+        Manager.scope_strategies[scope] = block.call current_strategies if block_given?
+      end
+    end
+
+    ##
+    # This module is only there to declare all used scopes. Technically a scope can be an
+    # arbitrary symbol. But we declare them here not to lose sight of them.
```
Comment (NobodysNightmare) [inline lib/open_project/authentication.rb:49]: Should we (by convention) refer to those as `OpenProject::Authentication::Scope::API_V3`?

I have to admit that it was not clear to me where `API_V3` originated from when it was used.

But you might also say that fully qualifying it is too verbose...

Comment (machisuji) [inline lib/open_project/authentication.rb:49]: I have used the respective constant but not fully qualified because of the verbosity, yes.
I suppose I'm not confused seeing just the constant as RubyMine tells me where it comes from.

File: lib/open_project/authentication.rb (hunk 5 above)
Comment (NobodysNightmare) [inline lib/open_project/authentication.rb]: I see... Would be a worthwhile mention, something brief like 

> unchanged if omitted, defaults to false

 or something similar comes to my mind here...

anyway: Wow there is documentation :+1:
That means my comment comes from a quite high satisfaction level :cake:

File: config/initializers/warden.rb (hunk 4 above)
Comment (machisuji) [inline config/initializers/warden.rb]: It used to be more than 1 value per key. But yeah, I'll go ahead and make it a dictionary.

File: lib/open_project/authentication.rb (hunk 6 above)
Comment (NobodysNightmare) [inline lib/open_project/authentication.rb:49]: Good thing... I am currently on GitHub-only mode ;-)

Think we can keep it for now... You wanted to re-cut the scopes "later" anyway, didn't you?

File: lib/open_project/authentication/manager.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
+module OpenProject
+  module Authentication
+    class Manager < Warden::Manager
+      serialize_into_session do |user|
```
Comment (NobodysNightmare) [inline lib/open_project/authentication/manager.rb:4]: @machisuji Would you agree to @houndci here?

edit: I think I don't agree...

File: lib/open_project/authentication/strategies/warden/basic_auth_failure.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+module OpenProject
+  module Authentication
+    module Strategies
+      module Warden
+        ##
+        # This strategy is inserted after optional basic auth strategies to
+        # indicate that invalid basic auth credentials were provided.
+        class BasicAuthFailure < ::Warden::Strategies::BasicAuth
+          def authenticate_user(_username, _password)
+            nil # always fails
```
Comment (NobodysNightmare) [inline lib/open_project/authentication/strategies/warden/basic_auth_failure.rb:10]: I feel stupid, but I have to ask: Why?

File: lib/open_project/authentication/strategies/warden/global_basic_auth.rb [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+          ##
+          # Only valid if global basic auth is configured and tried.
+          def valid?
+            self.class.configuration? && super && username == self.class.user
+          end
+
+          def authenticate_user(username, password)
+            if username == self.class.user && password == self.class.password
+              User.system.tap do |user|
+                user.admin = true
```
Comment (NobodysNightmare) [inline lib/open_project/authentication/strategies/warden/global_basic_auth.rb:79]: This feels dirty...

Let me guess: There was a time, when the system user was not an admin, but now we need him to be?

File: lib/open_project/authentication/strategies/warden/basic_auth_failure.rb (hunk 8 above)
Comment (machisuji) [inline lib/open_project/authentication/strategies/warden/basic_auth_failure.rb:10]: The `GlobalBasicAuth` and `UserBasicAuth` strategies are both optional. I.e. they are only triggered when certain user names are used for basic auth:

1) What ever is configured for the global basic auth access.
2) `apikey`

This means if a user provides a different user name the credentials would be wrong but the user would not get a 401 because the strategies were not triggered due to the different user name. Instead the session strategy will be used as a fallback and the user will be either logged in or anonymous.

But when they provide credentials they should get a 401 if they are wrong. This is what this strategy is for.
It always fails. But is only triggered when basic auth credentials are provided. If the credentials are correct, one of the previous two basic auth strategies will succeed. If not they must be wrong.

This wouldn't be necessary if the two basic auth strategies were always triggered. For instance if the `UserBasicAuth` strategy would work with usernames and passwords instead of `apikey`.

File: lib/open_project/authentication/strategies/warden/global_basic_auth.rb (hunk 9 above)
Comment (machisuji) [inline lib/open_project/authentication/strategies/warden/global_basic_auth.rb:79]: The SystemUser is apparently generally no admin. But the point of the `GlobalBasicAuth` strategy is to provide global admin access to an OpenProject instance **without requiring a user** - otherwise one could just use `UserBasicAuth` with a user who is an admin. (remember: in the com context no admin user will be seeded)

So using the System user seemed to make sense. Dunno why the system user is no admin to begin with.

File: lib/open_project/authentication/strategies/warden/session.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+          end
+
+          def authenticate!
+            user = user_id ? User.find(user_id) : User.anonymous
+
+            success! user
+          end
+
+          def user_id
+            Hash(session)['user_id']
```
Comment (NobodysNightmare) [inline lib/open_project/authentication/strategies/warden/session.rb:21]: I assume the trick here is, that this is not the place in the session where warden stores its information?

Meaning that this strategy gets the session information in the "old" way, while the warden store is a new way of doing it?

Comment (machisuji) [inline lib/open_project/authentication/strategies/warden/session.rb:21]: Correct! Later, when everything is unified it wouldn't work this way anymore. The authentication is transparent. You just use `warden.authenticate!` and warden gets the user from the session if available or runs the strategies to authenticate.

File: lib/open_project/authentication/strategies/warden/basic_auth_failure.rb (hunk 8 above)
Comment (NobodysNightmare) [inline lib/open_project/authentication/strategies/warden/basic_auth_failure.rb:10]: Okay, I think now I've got it. In my own words:

Whenever a stragegy is valid it decides _ultimately_ over grant or denial, that means the next strategy in the list is **only** tried when the current one was **not** valid.

This means without Basic auth headers, all (three) basic auth strategies are invalid, thus `session` has to decide.

With basic auth headers it depends:
At least the fail strategy is hit, so we don't fall back to session, because the fail strategy comes first (but only when HTTP basic auth headers are present).
In the special cases `user == 'apikey'` or `user == 'what_was_configured'` the other strategies are valid, but will decide _ultimately_ themselves.

That means **this** strategy only denies access when:
- HTTP basic auth is used
- none of the other basic auth strategies hit

That also means, if this strategy would be missing:
- fallback would be the session strategy (or no strategy under different configuration circumstances)
- only if session would recognize you, you would **not** get a 401
  - i.e. no backdoor would ever be introduced alone by forgetting this strategy (this is what I feared in the first place)

Comment (machisuji) [inline lib/open_project/authentication/strategies/warden/basic_auth_failure.rb:10]: 10/10 would confirm again.

File: lib/open_project/authentication/strategies/warden/global_basic_auth.rb (hunk 9 above)
Comment (NobodysNightmare) [inline lib/open_project/authentication/strategies/warden/global_basic_auth.rb:79]: The problem I have here is that the system user will now only be an admin, when he was authenticated using warden. Otherwise he is still no admin.

I don't know how I feel about that.

File: spec/lib/api/v3/authentication_spec.rb [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+require 'spec_helper'
+
+describe API::V3, type: :request do
+  describe 'basic auth' do
+    let(:user) { FactoryGirl.create :user }
+    let(:resource) { "/api/v3/users/#{user.id}"}
```
Comment (NobodysNightmare) [inline spec/lib/api/v3/authentication_spec.rb]: @houndci is telling the truth @machisuji

File: spec/lib/api/v3/authentication_spec.rb [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+require 'spec_helper'
+
+describe API::V3, type: :request do
```
Comment (NobodysNightmare) [inline spec/lib/api/v3/authentication_spec.rb]: Isn't this an integrative test that should reside under `spec/requests/api/v3`?

I also only grew into that structure that our integrative tests reside under `requests`... and this looks fitting here too

File: spec/lib/api/v3/authentication_spec.rb [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+      before do
+        Setting.login_required = 1
+      end
+
+      context 'with global basic auth configured' do
+        let(:username) { 'root' }
+        let(:password) { 'toor' }
+
+        before do
+          Strategies::GlobalBasicAuth.configure! user: 'root', password: 'toor'
```
Comment (NobodysNightmare) [inline spec/lib/api/v3/authentication_spec.rb]: you could re-use the username and password variables here...

Comment (machisuji) [inline spec/lib/api/v3/authentication_spec.rb]: No, I can't. If I do that the semantics of this before block change. It's supposed to always configure `root` and `toor` as the global basic auth credentials. Below in the nested user basic auth section `username` and `password` are changed, though, which would change that as well. Meaning the global basic auth strategy would match instead of the intended user basic auth strategy.

File: spec/lib/api/v3/authentication_spec.rb [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+        end
+
+        it_behaves_like 'it is basic auth protected'
+
+        describe 'user basic auth' do
```
Comment (NobodysNightmare) [inline spec/lib/api/v3/authentication_spec.rb]: I don't understand why this block is a child of the block `with global basic auth configured`

Comment (machisuji) [inline spec/lib/api/v3/authentication_spec.rb]: It's to test that both work in conjunction and that the global auth will be skipped if another username is used.

File: spec/lib/api/v3/authentication_spec.rb (hunk 13 above)
Comment (NobodysNightmare) [inline spec/lib/api/v3/authentication_spec.rb]: haha, which leads to my next question xD

Comment (NobodysNightmare) [inline spec/lib/api/v3/authentication_spec.rb]: ~~and btw it should also not match, because for the user basic auth you should be using `user == 'apikey'`, but let me check again...~~

File: spec/lib/api/v3/authentication_spec.rb (hunk 14 above)
Comment (NobodysNightmare) [inline spec/lib/api/v3/authentication_spec.rb]: This leads to those weird edge cases in testing where you never check that the user authentication works **without** global authentication...

**Comment by NobodysNightmare** [comment]
The server returns 0 Bytes on me for this request, need to investigate further:

```
GET http://localhost:5000/api/v3/work_packages/74 HTTP/1.1
Host: localhost:5000
Connection: keep-alive
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
Postman-Token: a0779441-a613-b6f3-3c1a-2181bbf75759
Cache-Control: no-cache
Authorization: Basic aGFuczp3dXJzdA==
Accept-Language: de-DE,de;q=0.8,en-US;q=0.6,en;q=0.4
User-Agent: Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36
Content-Type: application/json
Accept-Encoding: gzip, deflate, sdch

```

my console tells me:

```
!! Unexpected error while processing request: undefined method `each' for "unauthorized":String
```

edit: Happens for me with all kinds of wrong credentials, also for `apikey:wurst` and `apikey2:theWorkingRealApiKey` while `apikey:theWorkingRealApiKey` works as expected.

File: lib/open_project/authentication/manager.rb (hunk 7 above)
Comment (machisuji) [inline lib/open_project/authentication/manager.rb:4]: I don't agree either. With common use cases such as `#map` I agree. But not here even though it is essentially just a map too. With the short form it's not obvious what is being mapped, however.

**Comment by machisuji** [comment]
Ah that looks like a wrong rack error response then. I.e. `'unauthorized'` instead of `['unauthorized']`. I'll fix it.

**Comment by machisuji** [comment]
The question is why this does not occur in the specs.

**Comment by machisuji** [comment]
@NobodysNightmare I did two more things:

1) Implemented custom failure responses so API v3 can return JSON in all cases.
2) Made sure the WWW-Authenticate header is set upon authentication failure.

Not so sure if the latter is absolutely necessary but technically the HTTP spec ([10.4.2 401 Unauthorized](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html)) dictates that.

**Comment by Azure7111** [comment]
@machisuji Did you by any chance notice https://community.openproject.org/work_packages/18368 when addressing your JSON responses? I did not have an in depth look at it, but basically the usage of `::API::Errors::ErrorBase` might not always result in proper JSON.

**Comment by NobodysNightmare** [comment]
@Azure7111 I updated the ticket you linked. Note that this is a general issue we have in development mode.

@machisuji You broke the tests.

You are right that we were violating the specs here in not providing a `WWW-Authenticate`.
I will have to check what you did here (currently commenting without a look at the code or anything) and which pratical consequences this has...

**Comment by machisuji** [comment]
@NobodysNightmare It's all green now.

File: spec/requests/api/v3/authentication_spec.rb [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+          get resource
+        end
+
+        it 'should return 401 unauthorized' do
+          expect(response.status).to eq 401
+        end
+
+        it 'should return the correct JSON response' do
+          expect(JSON.parse(response.body)).to eq response_401
+        end
```
Comment (NobodysNightmare) [inline spec/requests/api/v3/authentication_spec.rb:66]: This + the 401 check should be replaceable with `it_behaves_like 'unauthenticated access'`.

The afforementioned shared example can be found in `spec/requests/api/v3/support/response_examples.rb` and uses `last_response` which is hopefully equivalent to `response` ^^

**Comment by NobodysNightmare** [comment]
```
POST http://localhost:5000/api/v3/work_packages/4/form
```

**Actual Response (your PR)**

> Cache-Control → no-cache
> Connection → keep-alive
> Content-Length → 170
> **Content-Type → text/html**
> Date → Thu, 21 May 2015 09:36:44 GMT
> Server → thin 1.5.1 codename Straight Razor
> WWW-Authenticate → Basic realm="private area"
> **X-Frame-Options → SAMEORIGIN**
> X-Rack-Cache → invalidate, pass
> X-Request-Id → 5ac65604ee6ebb65c148704c1c04f032
> X-Runtime → 0.589440
> X-UA-Compatible → IE=Edge,chrome=1
> 
> {"_type":"Error","errorIdentifier":"urn:openproject-org:api:v3:errors:MissingPermission","message":"You need to be authenticated to access this resource.","_embedded":{}}

**Expected Response (`dev` branch)**

> Cache-Control → no-cache
> Connection → keep-alive
> Content-Length → 164
> **Content-Type → application/hal+json; charset=utf-8**
> Date → Thu, 21 May 2015 09:34:13 GMT
> Server → thin 1.5.1 codename Straight Razor
> X-Rack-Cache → invalidate, pass
> X-Request-Id → 8af54c7e9b0c6464a030449694f51367
> X-Runtime → 0.901379
> X-UA-Compatible → IE=Edge,chrome=1
> 
> {"_type":"Error","errorIdentifier":"urn:openproject-org:api:v3:errors:MissingPermission","message":"You are not authorized to access this resource.","_embedded":{}}

The `Content-Type` is wrong on this branch.
I am not quite sure whether the `X-Frame-Options` are valid and neccessary here.

**Comment by NobodysNightmare** [comment]
RFC:
When I provide the wrong credentials via basic auth I get this response:

```
{
    "_type" : "Error",
    "errorIdentifier" : "urn:openproject-org:api:v3:errors:MissingPermission",
    "message" : "You need to be authenticated to access this resource.",
    "_embedded" : {}
}
```

I assume that the code rendering the error does know about the failed authentication, because I've never seen this error message before.

Following proposals:
- introduce a new `errorIdentifier` (incl. one-liner in api doc), because `MissingPermission` implies that I am successfully recognized as anyone
  - e.g. `urn:openproject-org:api:v3:errors:AuthenticationFailed`
- Make the error message more specific, because not only am I not authenticated, but I tried to do it and failed
  - e.g. "Authentication with the given credentials failed."

The first proposal is to allow better automated distinction of errors.
The second proposal is due to the fact that the `message` is designed to be immediately displayed to the user. That means he should be able to make sense out of the message and "... authentication failed ..." is more comprehensible than "... not authenticated ...".

However, note that I am currently not sure which information we have when generating the error message (e.g. do we know that authentication was attempted? do we know it was basic auth?)

**Comment by machisuji** [comment]
@NobodysNightmare [Fixed](https://github.com/machisuji/openproject/commit/c1e597c18615883b9407be92cc4e7b6821df769d) the content type also noticed that the realm was wrong. So fixed that too.

**Comment by machisuji** [comment]
We do know if authentication was attempted. We also know whether it was basic auth or not. Though we only have basic auth right now which could fail.

I've added your proposal including distinct error messages for missing and wrong credentials.
Though I called the error Unauthenticated like the respective class. If you prefer 'AuthenticationFailed', we can change that, though.

**Comment by NobodysNightmare** [comment]
> Though I called the error Unauthenticated like the respective class. If you prefer 'AuthenticationFailed', we can change that, though.

At first I wanted to agree, but now I am not sure whether we are thinking about the same semantics here.
Until now we had the same error identifier for both 401 and 403, because it does not really matter (RFC: does it?) whether your permissions were missing due to missing auth or due to missing permissions of an existing account.

I thought that it would semantically make sense to have a distinct code for "you tried hard to authenticate, but still failed". The error message provided fits that purpose, but I am not sure whether the error code is apropriate.

My question is: Would you only display `Unauthenticated` for a failed authentication or generally when the user has missing permissions, but could authenticate?

In the latter case I'd prefer the term `Unauthenticated`, but we would also need to change some more existing behaviour.

In the former case I'd prefer `AuthenticationFailed`.

P.S.: Just realized that your doc changes point to the latter opinion...

**Comment by machisuji** [comment]
From the sole source of truth Wikipedia:

> Status codes 401 (Unauthorized) and 403 (Forbidden) have distinct meanings.
> 
> A 401 response indicates that access to the resource is restricted, and the request did not provide any authentication. It is possible that a new request for the same resource will succeed if authentication is provided. That might be accomplished via an Authorization header, a login cookie, or in some other way.
> 
> A 403 response generally indicates one of two conditions:
> 
> Authentication was provided, but the authenticated user is not permitted to perform the requested operation.
> The operation is forbidden to all users. For example, requests for a directory listing return code 403 when directory listing has been disabled.

That's the semantics I'm thinking of. And the current behaviour (in the PR) should reflect that.
The specs should be fixed. It was just a single shared example failing. Travis will tell.

edit: ok the wikipedia section doesn't actually mention the case of provided but wrong credentials. But those fall into the first category (401) IMO.

**Comment by NobodysNightmare** [comment]
We can distinct between our `errorIdentifiers` and HTTP status codes. Otherwise we would not need them, because we already have the HTTP status code ^^

I agree that we should make spec conformant use of HTTP status codes, but that was not really my question ^^ My question was about our use of `errorIdentifiers`, which provide a second level of machine readable status indication.

**Comment by oliverguenther** [comment]
Just adding my 2 cents to the identifiers:

> @machisuji 
> edit: ok the wikipedia section doesn't actually mention the case of provided but wrong credentials. 
> But those fall into the first category (401) IMO.

As per rfc2617 (cf., [Sec 1.2](https://www.ietf.org/rfc/rfc2617.txt)), erroneous authentication (user/pass combination) attempts should return 401 with a new challenge.

IMHO, error handling of both invalid credentials with 401 and insufficient access with 403 would suffice for machine readability, as is the common approach for Basic Auth anyway.
However since OP's API exposes a second level of errors on top of HTTP codes, adding the Unauthenticated error for the former, and exploiting MissingPermission for the latter case would be consistent.

**Comment by machisuji** [comment]
Thanks for your input @oliverguenther. So everything's jolly good then, isn't it?

**Comment by oliverguenther** [comment]
I've just skimmed through the changes I've missed the past days and tested the feature.
Two remaining remarks concerning GlobalBasicAuth: 
1. Is that something that should be contained in the exemplary configuration.yml (commented, as in the class docstring)?
2. Due to the comparison in `GlobalBasicAuth#authenticate_user`, number-only passwords have to be enclosed in quotes, otherwise `==` will fail. (Found this as I simply copy-pasted that part from the docstring).  `configuration[:password].to_s` in `self.password` would fix it.

This works really well otherwise. :+1: :+1:   from me.

**Comment by machisuji** [comment]
Ah good catches! Cheers. Will fix it.

**Comment by NobodysNightmare** [comment]
@oliverguenther Thanks for your feedback. I would never have spotted that int vs string error (the days I look back at my statically typed times...)

@machisuji You fixed that yesterday? On a public holiday? This is what happens if you work on public holidays:

```
ArgumentError: password must not be empty
```

:D

**tl; dr:** Now I _have to_ configure global basic auth ^^

**Comment by NobodysNightmare** [comment]
I am doing another manual test run. Looking good so far.

One thing I noticed is kind of related, but not really:
To obtain your own API-Key, you have to go to your account settings (`/my/account`). However, the API-Key is only displayed when the instance setting "Enable REST web service" (aka `settings_rest_api_enabled`) is enabled. Otherwise the API-Key is hidden.

This made sense when the API-Key was only useful in conjunction with the APIv2, that could be disabled in some parts. However, as APIv3 can't be disabled we now need to always display the API-Key.

If you feel that such a change does not belong here, feel free to open a second PR. But I hope it is only a one or two line change ;)

edit:
After enabled the rest-service on teatro I read:

> API access key created less than a minute ago

I might have underestimated the effort for this task oO (I thought about a _single_ if statement, but maybe there is also some creation of keys involved...)

**Comment by NobodysNightmare** [comment]
Looking good,

I'll ignore the travis failure because your last changes could not have caused them :fire:
